Projects/Project10 Names to Faces/Project10 Names to Faces/ViewController.swift

```diff
@@ -7,56 +7,132 @@
 //
 
 import UIKit
+import LocalAuthentication
 
-class ViewController: UICollectionViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
+class ViewController: UICollectionViewController {
     
     var people = [Person]()
-
+    
+    var authenticateButton: UIBarButtonItem?
+    var addFromPhotosButton: UIBarButtonItem?
+    var addFromCameraButton: UIBarButtonItem?
```

Comment:
This can be written as lazy vars 

```swift 
lazy var authenticateButton: UIBarButtonItem = {
    return UIBarButtonItem(title: "Authenticate", style: .plain, target: self, action: #selector(authenticateTapped))
}()
```
this way they don't need to be optional and are only assigned the first time they are referenced.

Reply:
Okay Derek. I've read the documentation on properties but I'm getting stumped on the syntax you used. So I understand we are using the lazy keyword instead of the optional. I want to use lazy so that these buttons will get instantiated and assigned only once when the property is first used. Now my question is how you are assigning the block of code to the property. I'm a bit confused because it looks like a mixture between and closure and computed property but Im assuming its neither of them because you're using the assignment operator = and also the function operator ().  Am I over thinking this? Are you just assigning a function? And if so, why are you returning the instantiated object in a block of code? 
Whats the trade off of just simply using lazy var authenticateButton: UIBarButtonItem =  UIBarButtonItem(title: "Authenticate", style: .plain, target: self, action: #selector(authenticateTapped)). 

Alright thanks whenever you get a second. no rush

Reply:
I've been thinking about this on why I think it would be better to use the lazy variables instead of optional variables. Correct me if I'm wrong. Both ways work but I shouldn't be trying to create optionals unless a task specifically requires it me to use them. It seems like it just adds complexity to the code syntax which can affect clarity. I guess if I have the option to create properties and take advantage of the lazy keyword it makes it easier on managing code. The whole idea of optionals is that if a variable can have the possibility of have a nil value it should be an optional so in this case the app NEEDS these properties and therefor they should never be nil and optionals shouldn't be used.

What do you think? alright thanks

Reply:
In short, you are correct. Use optionals when they represent something that could be `nil`. Not because you didn't initialize them at init time of the parent object. 

### Lazy Vars 
https://medium.com/@abhimuralidharan/lazy-var-in-ios-swift-96c75cb8a13a
This article covers lazy vars pretty well. The syntax I used uses a closure. 
Imagine that it is a custom unit of code that gets fired when you first access the property. 

The cool thing with this approach is you only use the memory space if you actually use it. 

Why does this matter? 
Consider the following: 

Let's say that each one of these buttons 50MB of memory and you have only 200MB total to use ( forget your actual use case)

```swift 
    var authenticateButton: UIBarButtonItem?
    var addFromPhotosButton: UIBarButtonItem?
    var addFromCameraButton: UIBarButtonItem?
 ``` 
If you were to initialize them all when you declare them, which means they get initialized when their parent object gets initialized see below: 

```swift 
var authenticateButton: UIBarButtonItem = UIBarButtonItem(title: "Authenticate", style: .plain, target: self, action: #selector(authenticateTapped))
var addFromPhotosButton: UIBarButtonItem = ... 
var addFromCameraButton: UIBarButtonItem = ... 
 ```
You would automatically be tying up 150 MB of memory. That would leave you with 50MB left and that's not a lot. 

Now let's add that you don't know if all of the buttons are going to be used. And some may be accessed only if a user does something that invokes it... 

Lazy vars will allow you to only create the button if you run code that accesses that var 

Which means when the view controller loads, you still have all 200MB. And only when you call the code to add the buttons, will the buttons actually be created and the memory be used. 

If you never call that func, the buttons never get created and you have your memory. 

This keeps a lot of the setup code contained in the var itself. And it optimizes memory management. 

But there is one catch. You cannot assign it to another value after you access it.

In the example above once you call the code to add the buttons, there is no way to assign nil, to free up the memory. 

In this case, you WILL want to make them optional because you do want some of them to be nil so you can free up space if you need to. 


But what about me just doing what I originally did? 

There technically is nothing wrong with what you are doing, but your use case doesn't suggest the need for them.

Reply:
Hi Derek. Okay okay I get a better picture of the pros and cons. That makes a lot of sense actually.
 I saw the way you implemented and I intend to make the change like you showed but I held off on it actually because I had a question regarding the syntax. I don't understand the implementation.

My question is how you are assigning the block of code to the property. I'm a bit confused because it looks like a mixture between a closure or computed property but I'm assuming its neither of them because you're using the assignment operator = and also the function operator ().
Am I over thinking this? Are you just assigning a function? And if so, why are you returning the instantiated object in a block of code instead of
**lazy var authenticateButton = UIBarButtonItem(title: "Authenticate", style: .plain, target: self, action: #selector(authenticateTapped)).**

thanks!

Reply:
Thanks a lot for the article @derek-bertubin . I did some digging around I understand now why you implemented it the way you did. It's a closure that gets executed lazily similar to a computer property but only once.